cloudformation/ad/ad-integration.yaml

```diff
@@ -507,22 +537,237 @@ Resources:
+      Handler: index.handler
+      MemorySize: 128
+      Role: !GetAtt DomainCertificateSetupLambdaRole.Arn
+      Runtime: python3.9
+      Timeout: 300
+      TracingConfig:
+        Mode: Active
+      Code:
+        ZipFile: |
+          import time
+          import cfnresponse
+          import boto3
+          import logging
+          import random
+          import string
+          logger = logging.getLogger()
+          logger.setLevel(logging.INFO)
+          acm = boto3.client("acm")
+          sm = boto3.client("secretsmanager")
+
+          def create_physical_resource_id():
+              alnum = string.ascii_uppercase + string.ascii_lowercase + string.digits
+              return ''.join(random.choice(alnum) for _ in range(16))
+
+          def handler(event, context):
+              redact_keys = {}
+              print('Context: {}'.format(context))
+              print({k:event[k] if k not in redact_keys else "[REDACTED]" for k in event.keys()})
+              print('boto version {}'.format(boto3.__version__))
+              domain_name = event['ResourceProperties']['DomainName']
+              domain_certificate_secret_arn = event['ResourceProperties']['DomainCertificateSecretArn']
+              domain_private_key_secret_arn = event['ResourceProperties']['DomainPrivateKeySecretArn']
+              acm_certificate_secret_arn = event['ResourceProperties']['ACMCertificateSecretArn']
+
+              response_data = {}
+              reason = None
+              response_status = cfnresponse.SUCCESS
+
+              try:
+                if event['RequestType'] == 'Create':
+                  response_data['Message'] = 'Resource creation successful!'
+                  physical_resource_id = create_physical_resource_id()
+                  logger.info('Reading secrets from Secrets Manager...')
+                  domain_certificate = sm.get_secret_value(SecretId=domain_certificate_secret_arn)["SecretString"]
+                  domain_private_key = sm.get_secret_value(SecretId=domain_private_key_secret_arn)["SecretString"]
+                  logger.info('Importing certificate into ACM...')
+                  certificate_arn = acm.import_certificate(Certificate=domain_certificate, PrivateKey=domain_private_key)["CertificateArn"]
+                  logger.info('Storing ACM certificate ARN into Secrets Manager...')
+                  sm.put_secret_value(SecretId=acm_certificate_secret_arn, SecretString=certificate_arn)
+                  response_data['DomainCertificateArn'] = certificate_arn
+                elif event['RequestType'] == 'Delete':
+                  physical_resource_id = event['PhysicalResourceId']
+                  certificate_arn = sm.get_secret_value(SecretId=acm_certificate_secret_arn)["SecretString"]
+                  logger.info('Deleting ACM certificate {}...'.format(certificate_arn))
+                  acm.delete_certificate(CertificateArn=certificate_arn)
```

Comment:
This call may fail due to ACM certificate still being in use. We must implement a retry